### ISSUE TITLE
Map does not work as expected

### ISSUE THREAD

**Issue by pauhull** [jmattheis/goverter#105]
Take the following code:
```go
package example

// goverter:converter
type Converter interface {
	// goverter:map PropertyA PropertyB
	AToB(A) B
}

type A struct {
	PropertyA string
}

type B struct {
	PropertyB string
}
```

This works as expected and the field is correctly mapped. However, if either the source or the target struct (but not both) have a pointer receiver, the generation fails. For example, for
```go
// goverter:converter
type Converter interface {
	// goverter:map PropertyA PropertyB
	AToB(A) *B
}
```
the error would look as follows:
```
Error while creating converter method:
    func (example.Converter).AToB(example.A) *example.B

| example.A
|
|     | example.A
|     |
source*.???
target*.PropertyB
|     | |
|     | | string
|     |
|     | example.B
|
| *example.B

Cannot match the target field with the source entry: "PropertyB" does not exist.
```

**Comment by jmattheis** [comment]
Yeah, I wish I had declined the ux feature of supporting goverter:map on struct pointers. There are just so many edge-cases :D. Anyway, the problem is with a definition like this:

    type Converter interface {
        // goverter:map Name FullName
        Convert(source Input) *Output
    }

that goverter will currently create a internal submethod for the non pointer -> non pointer conversion. E.g.

    func (c *ConverterImpl) Convert(source execution.Input) *execution.Output {}
    func (c *ConverterImpl) structsInputToStructsOutput(source execution.Input) execution.Output {}

the `goverter:*` settings are only applied to stuff that's generated in the method it's declared on. If you'd declare it like this it'd work fine:

```
// goverter:converter
type Converter interface {
	// goverter:map PropertyA PropertyB
	AToB2(A) B
        AToB(A) *B
}
```
---

Another more hidden edge-case would be this:
```go
// goverter:converter
type Converter interface {
    Convert2(Input2) Output2

    // goverter:map FullName Name
    Convert(source Input) *Output
}
type Input struct { FullName string }
type Input2 struct { Value Input }
type Output2 struct { Value Output }
type Output struct { Name string }
```
The method `Convert2(Input2) Output2` will create the sub method `internalName(Input) Output`. Which won't use the settings defined on `Convert(source Input) *Output` and will then fail because the Name property is unmapped.

So to properly fix this bug, goverter has to not create a sub method for your use-case and fail with a good error message when goverter would create a sub method for a conversion when there is an explicit method definition with config that is similar to the internally created method. 

Similar to how it's done for when there is an overlap for explicit methods. E.g:
```go
// goverter:converter
type Converter interface {
    // goverter:ignore Age
    Convert(source *Input) (*Output)
    ConvertSub(source Input) (Output)
}
```
which already fails with:
```
    Overlapping field mapping found.

    Please move the field related goverter:* comments from this method:
        func (github.com/jmattheis/goverter/execution.Converter).Convert(source *github.com/jmattheis/goverter/execution.Input) *github.com/jmattheis/goverter/execution.Output

    to this method:
        func (github.com/jmattheis/goverter/execution.Converter).ConvertSub(source github.com/jmattheis/goverter/execution.Input) github.com/jmattheis/goverter/execution.Output

    Goverter will use ConvertSub inside the implementation of Convert, thus, field related settings would be ignored.

```

**Comment by jmattheis** [comment]
Could you try out the `map-on-pointer` branch and see if it works for your use-case? You can see the changes here: https://github.com/jmattheis/goverter/compare/map-on-pointer?expand=1

With `go install`
```
go install github.com/jmattheis/goverter/cmd/goverter@map-on-pointer
```
or `go run`
```
go run github.com/jmattheis/goverter/cmd/goverter@map-on-pointer [pattern]
```
or `go get`
```
go get github.com/jmattheis/goverter@map-on-pointer
```

**Comment by pauhull** [comment]
Thanks for the quick response and the in-depth explanation 🙂
I tried the new branch and `AToB(A) *B` seems to work now. However, `AToB(*A) B` still fails to generate:
```
Error while creating converter method:
    func (example.Converter).AToB(*example.A) example.B

| *example.A
|
source
target
|
| example.B

TypeMismatch: Cannot convert *example.A to example.B
```

**Comment by jmattheis** [comment]
This is by design. Goverter doesn't know what to do when the `source` is `nil`, because such a value cannot be represented in the not nullable `target`. You can enable [`useZeroValueOnPointerInconsistency`](https://goverter.jmattheis.de/#/config/useZeroValueOnPointerInconsistency) to instruct goverter to use the zero value of target if the `source` is `nil` or write a custom [`extend`](https://goverter.jmattheis.de/#/config/extend) method for this type combination.

I think it's time to improve the error message for this specific error, as it's not so obvious.

**Comment by pauhull** [comment]
Oh, that's good to know! I actually have that flag set in my main project, I just didn't include it in the example I wrote. Thanks for the clarification! 🙂

**Comment by jmattheis** [comment]
Fixed with [v1.2.0](https://github.com/jmattheis/goverter/releases/tag/v1.2.0).
